espnet/nets/pytorch_backend/e2e_st.py

```diff
@@ -266,56 +291,87 @@ def forward(self, xs_pad, ilens, ys_pad, ys_pad_asr):
         # 1. Encoder
         hs_pad, hlens, _ = self.enc(xs_pad, ilens)
 
-        # 2. ASR loss
-        if self.asr_weight == 0:
+        # 2. ST attention loss
+        self.loss_st, acc, _ = self.dec(hs_pad, hlens, ys_pad, lang_ids=tgt_lang_ids)
+        self.acc = acc
+
+        # 2. ASR CTC loss
+        if self.asr_weight == 0 or self.mtlalpha == 0:
+            self.loss_ctc = None
+        else:
+            self.loss_ctc = self.ctc(hs_pad, hlens, ys_pad_asr)
+
+        # 3. ASR attention loss
+        if self.asr_weight == 0 or self.mtlalpha == 1:
             self.loss_asr = None
             self.acc_asr = 0.0
         else:
             self.loss_asr, acc_asr, _ = self.dec_asr(hs_pad, hlens, ys_pad_asr)
             self.acc_asr = acc_asr
 
-        # 3. ST loss
-        self.loss_st, acc, _ = self.dec(hs_pad, hlens, ys_pad, tgt_lang_ids=tgt_lang_ids)
-        self.acc = acc
+        # 4. compute cer without beam search
+        if (self.asr_weight == 0 or self.mtlalpha == 0) or self.char_list is None:
+            cer_ctc = None
+        else:
+            cers = []
+
+            y_hats = self.ctc.argmax(hs_pad).data
+            for i, y in enumerate(y_hats):
+                y_hat = [x[0] for x in groupby(y)]
+                y_true = ys_pad_asr[i]
 
-        # 4. compute cer/wer
-        if self.asr_weight == 0:
+                seq_hat = [self.char_list[int(idx)] for idx in y_hat if int(idx) != -1]
```

Comment:
Probably, you could try with https://github.com/espnet/espnet/blob/f10be2416bf14bdc33d00313b95d2944f954ecc5/espnet/nets/e2e_asr_common.py#L88

Reply:
@Fhrozen You are right. But the RNN model has a beam search option during training while Transformer allows greedy decoding only. That's why these interfaces for the WER calculation are different. Maybe it's better to integrate them in another PR.

Reply:
I see.